File: c/Arquivo.c
#include <stdio.h>
#include <stdlib.h>

int main() {
    FILE *arquivo;
    int n;
    scanf("%d", &n);

    // Abertura do arquivo para escrita
    arquivo = fopen("valores.txt", "wb");
    if (arquivo == NULL) {
        printf("Erro ao abrir o arquivo para escrita.\n");
        return 1;
    }

    // Leitura dos valores reais e escrita no arquivo
    double valor;
    for (int i = 0; i < n; i++) {
        scanf("%lf", &valor);
        fwrite(&valor, sizeof(double), 1, arquivo);
    }

    // Fechar o arquivo
    fclose(arquivo);

    // Reabrir o arquivo para leitura de trás para frente
    arquivo = fopen("valores.txt", "rb");
    if (arquivo == NULL) {
        printf("Erro ao abrir o arquivo para leitura.\n");
        return 1;
    }

    // Ler e mostrar os valores de trás para frente
    for (int i = n - 1; i >= 0; i--) {
        fseek(arquivo, i * sizeof(double), SEEK_SET);
        fread(&valor, sizeof(double), 1, arquivo);
        if (valor == (int)valor) {
            printf("%d\n", (int)valor); // Imprimir como inteiro
        } else {
            printf("%.3lf\n", valor); // Caso contrário, imprimir com a parte decimal
        }
    }

    // Fechar o arquivo novamente
    fclose(arquivo);

    return 0;
}

File: java/Arquivo.java
import java.io.*;

public class Arquivo {
    public static void main(String[] args) {
        try {
            // Leitura do tam
            BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
            int n = Integer.parseInt(br.readLine());

            // Abertura do arquivo para escrita
            RandomAccessFile arquivo = new RandomAccessFile("valores.txt", "rw");

            // Leitura dos valores reais e escrita no arquivo
            for (int i = 0; i < n; i++) {
                double valor = Double.parseDouble(br.readLine());
                arquivo.writeDouble(valor);
            }

            // Fechar o arquivo
            arquivo.close();

            // Reabrir o arquivo para leitura de trás para frente
            arquivo = new RandomAccessFile("valores.txt", "r");

            // Ler e mostrar os valores de trás para frente
            for (int i = n - 1; i >= 0; i--) {
                arquivo.seek(i * 8); // Tamanho de um double em bytes
                double valor = arquivo.readDouble();
                // Verificar se o valor é um número inteiro
                if (valor == (int) valor) {
                    System.out.println((int) valor); // Imprimir como inteiro
                } else {
                    System.out.println(valor); // Caso contrário, imprimir com a parte decimal
                }
            }
            arquivo.close();
        } catch (IOException e) {
            e.printStackTrace();
        } catch (NumberFormatException e) {
            System.out.println("Erro na leitura de valores reais.");
        }
    }
}
